Add RAISE WARNING (NOTICE, INFO, etc) statement

andrioni opened the pull request:
This is a quick proposal for adding a new kind of statement, `RAISE WARNING[|NOTICE|INFO|LOG|DEBUG]`, inspired by the [PL/pgSQL RAISE statement](https://www.postgresql.org/docs/current/plpgsql-errors-and-messages.html), that is able to send informational messages to the client.

I'm not intending for this to be publicly documented, but to be used internally as an easy way to send WARNING/NOTICE/LOG/INFO/DEBUG messages to clients to do #10839. This statement is only available under experimental mode, with no plans to remove that restriction.

This PR is now in its final state, with all intended features implemented.

### Motivation

  * This PR helps with getting #10839 done.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

benesch reviewed (commented):
I'm a bit nervous about adding a PL/pgSQL function to the grammar directly. (Just haven't thought through the implications if one day we want to layer on a PL/pgSQL, but our SQL parser already accepts some arbitrary subset of PL/pgSQL commands.) Can you live with this being experimental? +1 to the idea if so.

andrioni commented:
Sure, no problems at all with keeping this under experimental forever, especially since it's mostly to help us test clients/drivers. It doesn't have anything very specific to `PL/pgSQL`, it's just that it's mostly useless otherwise, since SQL UDFs can only have a single statement. I guess you could use it inside stored procedures, but raw SQL procs are also pretty rare in my experience.

sploiselle commented on src/sql-parser/src/ast/defs/statement.rs:
```diff
@@ -1760,3 +1762,14 @@ impl AstDisplay for DeallocateStatement {
+
+/// `RAISE ...`
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct RaiseWarningStatement {}
```
Empty structs don't need a "body":

```diff
diff --git a/src/sql-parser/src/ast/defs/statement.rs b/src/sql-parser/src/ast/defs/statement.rs
index 2d6cbbb84..32f345c61 100644
--- a/src/sql-parser/src/ast/defs/statement.rs
+++ b/src/sql-parser/src/ast/defs/statement.rs
@@ -1765,7 +1765,7 @@ impl_display!(DeallocateStatement);
 
 /// `RAISE ...`
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
-pub struct RaiseWarningStatement {}
+pub struct RaiseWarningStatement;
 
 impl AstDisplay for RaiseWarningStatement {
     fn fmt<W: fmt::Write>(&self, f: &mut AstFormatter<W>) {
diff --git a/src/sql-parser/src/parser.rs b/src/sql-parser/src/parser.rs
index 301eb145a..b5b810f09 100644
--- a/src/sql-parser/src/parser.rs
+++ b/src/sql-parser/src/parser.rs
@@ -4479,7 +4479,7 @@ impl<'a> Parser<'a> {
     /// has already been consumed.
     fn parse_raise(&mut self) -> Result<Statement<Raw>, ParserError> {
         self.expect_keyword(WARNING)?;
-        Ok(Statement::RaiseWarning(RaiseWarningStatement {}))
+        Ok(Statement::RaiseWarning(RaiseWarningStatement))
     }
 }
```

benesch replied:
IIUC this will become `Statement::Raise` with a `severity: Severity` field?

andrioni replied:
Almost certainly yes, because I don't want to write five different statements, I just did it this way to get this draft out sooner.

andrioni replied:
Unless having five different statements is better! I'm going to have to use (yet another) custom severity enum here, since only the five values mentioned in the PR are accepted by the Postgres equivalent. That means no Error, Fatal or Panic, which seems fine to me, since right now this statement doesn't not have error semantics (e.g it doesn't abort open transactions).

sploiselle replied:
Ah, I misunderstood to what degree this is a WIP and that the severity fields were definitely going in before this was ready for a more thorough review. Apologies.

andrioni replied:
Sorry, I should have made that clearer!

andrioni replied:
Done!

sploiselle reviewed (commented):
I'm headed into the office so will look at this in a bit, but if Nikhil Oks w/ experimental mode, I'm sure that's a sane approach

mjibson commented:
I'm interested in seeing more motivation for this, even as an experimental feature. I currently don't see the need for a top level statement that does this, I thought we needed to improve other internals to allow for more places for these sorts of responses to be generated. How will this feature help #10839?

andrioni commented:
>How will this feature help https://github.com/MaterializeInc/materialize/issues/10839?

It actually allows you to generate NOTICE/INFO/LOG/WARNING/DEBUG messages from Materialize to see how they behave on each driver/client (e.g. if you are using Sequelize, can you even read these messages? Does it work by attaching a handler callback or is it stored somewhere in your connection object?), so that we can document it and use it to decide if we want to invest more in actually adding useful NOTICE-like messages (in case they are accessible/useful for most drivers/clients) or try to explore other alternatives of dealing with user notifications.

Right now it's impossible to generate most of these messages from SQL, as there are no codepaths that actually lead to pushing out a NOTICE or an INFO. These statements provide an easy way for that.

In more concrete terms, [DevEx](https://github.com/MaterializeInc/developer-experience) (#5187) is currently testing several drivers/frameworks and documenting our compatibility/support levels, and this provides them with an easy way to also cover #10839 in their efforts.

>I currently don't see the need for a top level statement that does this

I went this way mostly because it seemed to be the simplest and most "elegant" way of doing it in our codebase right now: `PlanError`s get coerced along the way to `anyhow::Error`s, which makes it harder to have the information in pgwire, and in the codepath processing `EvalError`s we have some assertions that they are in fact errors and not lower severity, which makes working with those a bit hackier. Happy to tackle it in a different way, I just don't have any other ideas right now on how to do it.

mjibson commented:
Ok so it's syntax mostly to allow us to easily test things, but probably not for users? If that's the case then yes let's do it.

andrioni commented:
Yep, exactly! I don't expect any users to ever use it, since it's mostly pointless for them.

andrioni commented:
Hey @benesch @sploiselle @mjibson, quick ping that I've now converted this into a proper PR, with the following main changes:
- All notice severity levels are now implemented (from DEBUG to WARNING)
- Required experimental mode
- Added pgtest

benesch reviewed (approved):
This looks wonderful!

andrioni commented:
@mjibson I'm switching to run the pgtests under experimental mode by default, any objections to that?

sploiselle commented on src/sql-parser/src/parser.rs:
```diff
@@ -4473,6 +4474,23 @@ impl<'a> Parser<'a> {
+            Some(LOG) => NoticeSeverity::Log,
+            Some(NOTICE) => NoticeSeverity::Notice,
+            Some(WARNING) => NoticeSeverity::Warning,
+            Some(_) | None => {
```
nit: `Some(_)` is unreachable, not a client error

andrioni replied:
Hi, I did a bit more of testing, and changing it to unreadable causes statements such as `RAISE BEGIN;` to panic, so I'll keep it as it is.

sploiselle replied:
That sounds like there's a bug in the parser then; give me a bit to look into this

andrioni replied:
Ah, sorry, I messed up while testing this, `Some(_)` is indeed unreachable, I just typoed when changing and also mapped `None` to unreachable.

sploiselle commented on src/materialized/tests/pgwire.rs:
```diff
@@ -427,7 +427,7 @@ fn test_record_types() -> Result<(), Box<dyn Error>> {
     // We want a new server per file, so we can't use pgtest::walk.
     datadriven::walk(dir.to_str().unwrap(), |tf| {
-        let server = util::start_server(util::Config::default()).unwrap();
+        let server = util::start_server(util::Config::default().experimental_mode()).unwrap();
```
I've had to do something akin to this when e.g. determining the output types of functions; seems reasonable for tests to run in experimental mode all the time and unclear what the downside would be.

sploiselle commented on src/pgwire/src/protocol.rs:
```diff
@@ -1262,6 +1262,27 @@ where
+            ExecuteResponse::Raise { severity } => {
+                let msg = match severity {
+                    NoticeSeverity::Debug => {
+                        ErrorResponse::debug(SqlState::WARNING, "raised a test debug")
```
Not very familiar with this area of the code, so apologies for the naive question, but the `ErrorResponse`'s text field doesn't show up anywhere; where does/should this get propagated?

andrioni replied:
You need to `set client_min_messages to debug` before! By default, we filter out messages that are lower severity than `NOTICE` (following Postgres behavior). If you check `raise.pt`, we set it before running the tests exactly because otherwise they would be no-ops from the client-side.

sploiselle replied:
Ah, I guess I'm confused--is this message supposed to be the one we want to commit to the repo, or is it a placeholder until we implement a use of this statement?

andrioni replied:
It's the message I want to commit, mostly because the existence of the message here is more important than the concrete message! The idea is to test how test how clients/drivers react to such messages, so the contents are more or less arbitrary.
